lib/identity/client.js

```diff
@@ -606,7 +618,117 @@ class Client extends PartialClient {
+  }
+  /**
+   * addSecretToNamespace takes a user defined namespace name and a secret to add to the namespace
+   *
+   * @param {String} namespace The user given name for the secret stored in the meta data
+   * @param {String} secretType The secret type, example 'Credential'
+   * @param {String} secretName The user given name for the secret stored in the meta data
+   *
+   * @return {Boolean}
+   */
+  async addSecretToNamespace(secretName, secretType, namespace) {
+    // Look up the user defined namespace
+    const groupName = `tozny.secret.${this.config.realmName}.${namespace}`
+    const groupList = await this.storage.listGroups(
+      this.storage.config.clientId,
+      [groupName]
+    )
+    let currentGroup
+    if (groupList.groups.length < 1) {
+      // Namespace Does not Exists, create one
+      const newGroup = await this.storage.createGroup(groupName, [], '')
+      currentGroup = newGroup.group
+      let groupMembers = []
+      const capabilities = {
+        read: true,
+        share: true,
+        manage: true,
+      }
+      const owner = new GroupMember(this.storage.config.clientId, capabilities)
+      groupMembers.push(owner)
+      await this.storage.addGroupMembers(currentGroup.groupID, groupMembers)
+    } else {
+      // Group Exists, save the group object
+      currentGroup = groupList.groups[0]
+    }
+    // Generate the Record Type to share
+    const recordType = `tozny.secret.${SECRET_UUID}.${secretType}.${secretName}`
+    let sharedWithGroup = await this.storage.shareRecordWithGroup(
+      currentGroup.groupID,
+      recordType
+    )
+    return sharedWithGroup.record_type
   }
+  /**
+   * addIdentityToNamespace takes a user defined namespace name and a username for an identity and gives access to the given namespace for the identity
+   *
+   * @param {String} namespace The user given name for the secret stored in the meta data
+   * @param {String} usernameToAdd The username of the user that will have permissions added
+   *
+   * @return {Boolean}
+   */
+  async addIdentityToNamespace(usernameToAdd, namespace) {
+    // Validation
+    if (usernameToAdd == '') {
+      throw new Error('Username Required')
+    }
+    // Look Up the ClientID for Username
+    let clientResponse = await this.searchRealmIdentitiesByUsername([
+      usernameToAdd,
+    ])
+    let usernameLowerCase = usernameToAdd.toLowerCase()
+    let clientID
+    if (clientResponse.searched_identities_information === null) {
+      return false
+    }
+    for (
+      let index = 0;
+      index < clientResponse.searched_identities_information.length;
+      index++
+    ) {
+      if (
+        clientResponse.searched_identities_information[index].realm_username ===
+        usernameLowerCase
+      ) {
+        clientID =
+          clientResponse.searched_identities_information[index].client_id
+      }
+    }
+    if (clientID == null) {
+      return false
+    }
+    // Check if namespace exists
```

Comment:
this logic of checking for if the namespace exists and if not creating it is very similar to the above, please extract out what is common to a helper function, and then call that in both places, e.g.  `let group = getOrCreateNamespace(namespace)`